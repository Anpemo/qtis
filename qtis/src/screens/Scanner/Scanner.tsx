import React, { useEffect } from 'react'
import { Text, View, StyleSheet, Button, Alert } from 'react-native'
import { BarCodeScanner } from 'expo-barcode-scanner'
import { connect } from 'react-redux'
import { AnyAction, bindActionCreators, Dispatch } from 'redux'
import { fetchProduct, cleanProduct } from '../../redux/actions/qtisActionCreators'
import ProductInterface from '../../Interfaces/productInterface'

const styles = StyleSheet.create({
  container: {
    flex: 1
  }
})

function Scanner ({ actions, product, navigation }: { actions: AnyAction, product: ProductInterface, navigation: any }) {
  const [scanned, setScanned] = React.useState(false)
  const [hasPermission, setHasPermission] = React.useState < boolean | null >(null)
  const [productBarCode, setProductBarCode] = React.useState < number | null >(null)

  useEffect(() => {
    (async () => {
      const { status } = await BarCodeScanner.requestPermissionsAsync()
      setHasPermission(status === 'granted')
    })()
  }, [])

  const handleBarCodeScanned = ({ data }: any) => {
    setScanned(true)
    if (data) {
      actions.fetchProduct(data.toString())
    }
    setProductBarCode(data)
  }
  function navigateToAddProduct () {
    setScanned(false)
    navigation.navigate('AddProduct', productBarCode)
    setProductBarCode(null)
  }
  useEffect(() => {
    if (scanned && product?.productBarCode) {
      navigation.navigate('ProductDetail', { productBarCode: product.productBarCode })
      setProductBarCode(null)
    } else if (scanned) {
      Alert.alert(
        'Product does not exist',
        'Do you want to add it?',
        [
          {
            text: 'Cancel',
            style: 'cancel'
          },
          {
            text: 'OK',
            onPress: () => navigateToAddProduct()
          }
        ],
        {
          cancelable: true
        }
      )
    }
  }, [product])

  if (hasPermission === null) {
    return <Text testID={'requesting'}>Requesting for camera permission</Text>
  }
  if (hasPermission === false) {
    return <Text testID={'camera'}>No access to camera</Text>
  }
  function resetScanner () {
    setScanned(false)
    actions.cleanProduct()
  }
  return (
    <View style={styles.container} >
      <BarCodeScanner
        onBarCodeScanned={scanned ? undefined : handleBarCodeScanned}
        style={StyleSheet.absoluteFillObject}
        testID={'scanner'}
      />

      {scanned && <Button title={'Tap to Scan Again'} onPress={() => resetScanner()} />}
    </View>
  )
}
function mapStateToProps ({ productsReducer }: Object) {
  return {
    product: productsReducer.product
  }
}

function mapDispatchToProps (dispatch: Dispatch) {
  return {
    actions: bindActionCreators({
      fetchProduct,
      cleanProduct
    }, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(Scanner)
