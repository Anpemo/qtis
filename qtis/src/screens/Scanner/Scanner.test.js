import { act, cleanup, fireEvent, render } from '@testing-library/react-native'
import Scanner from './Scanner'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { Alert } from 'react-native'

jest.mock('../../redux/actions/qtisActionCreators')

describe('Given a Scanner component', () => {
  let componentBarCode
  const navigate = jest.fn()
  beforeEach(() => {
    const mockStore = configureStore()
    componentBarCode = (
          <Provider store={mockStore({
            productsReducer: {
              product: { productBarCode: 5412 }
            }
          })}><Scanner navigation={{ navigate }}/></Provider>
    )
  })
  afterEach(() => {
    cleanup()
    jest.resetAllMocks()
  })
  describe('When rendering it', () => {
    test('Then scannerView will match the snapshot', () => {
      const scannerComponent = render(componentBarCode)
      expect(scannerComponent).toMatchSnapshot()
    })
    test('Then if hasPermission is false will render camera', () => {
      const reactUseState = React.useState
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => reactUseState(true))
        .mockImplementationOnce(() => reactUseState(false))

      const scannerComponent = render(componentBarCode)
      const cameraText = scannerComponent.getByTestId('camera')
      expect(cameraText).toBeTruthy()
    })
  })
})

describe('Given a Scanner component', () => {
  let componentNoBarCode
  const navigate = jest.fn()
  beforeEach(() => {
    const mockStore = configureStore()
    componentNoBarCode = (
            <Provider store={mockStore({
              productsReducer: {
                product: { }
              }
            })}><Scanner navigation={{ navigate }}/></Provider>
    )
  })
  afterEach(() => {
    cleanup()
  })
  describe('When rendering it', () => {
    test('Then if scanned is true will Alert', () => {
      const reactUseState = React.useState
      jest.spyOn(React, 'useState')
        .mockImplementationOnce(() => reactUseState(true))
      const { getByTestId } = render(componentNoBarCode)
      const scanButton = getByTestId('scanAgainButton')
      act(() => {
        fireEvent.press(scanButton)
      })
      jest.spyOn(Alert, 'alert')
      expect(scanButton).toBeTruthy()
    })
  })
})
