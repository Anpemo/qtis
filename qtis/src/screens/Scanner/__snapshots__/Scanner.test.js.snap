// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Scanner component When rendering it Then scannerView will match the snapshot 1`] = `
<Text
  testID="requesting"
>
  Requesting for camera permission
</Text>
`;
